packages/nimble-components/src/table/models/keyboard-navigation-manager.ts

```diff
@@ -0,0 +1,1025 @@
+import { tableHeaderTag } from '../components/header';
+import { MenuItem } from '../../menu-item';
+import { Menu } from '../../menu';
+import type { TabIndexOverride } from '../../patterns/tab-index-override/types';
+import { TableCellView } from '../../table-column/base/cell-view';
+
+interface TableFocusState {
+    focusType: TableFocusType;
+    headerActionIndex?: number;
+    rowIndex?: number;
+    columnIndex?: number;
+    cellContentIndex?: number;
+}
+
+interface TableHeaderFocusableElements {
+    headerActions: HTMLElement[];
+    columnHeaders: HTMLElement[];
+}
+
+/**
+ * Manages the keyboard navigation and focus within the table.
+ * @internal
+ */
+export class KeyboardNavigationManager<TData extends TableRecord>
+implements Subscriber {
+    private readonly focusState: TableFocusState = {
+        focusType: TableFocusType.none
+    };
+
+    private readonly tableNotifier: Notifier;
+    private focusableHeaderElements!: TableHeaderFocusableElements;
+    private inNavigationMode = true;
+    private handlingTabPress = false;
+
+    public constructor(
+        private readonly table: Table<TData>,
+        private readonly virtualizer: Virtualizer<TData>
+    ) {
+        table.addEventListener('keydown', e => this.onCaptureKeyDown(e), {
+            capture: true
+        });
+        table.addEventListener('keydown', e => this.onKeyDown(e));
+        table.addEventListener('focusin', e => this.handleFocus(e));
+        table.addEventListener('focusout', e => {
+            console.log(
+                'table focusout',
+                'target',
+                e.target,
+                'relatedTarget',
+                e.relatedTarget,
+                'active element',
+                this.getActiveElementDebug()
+            );
+        });
+        table.addEventListener('blur', e => {
+            console.log(
+                'table blur',
+                'target',
+                e.target,
+                'relatedTarget',
+                e.relatedTarget,
+                'active element',
+                this.getActiveElementDebug()
+            );
+        });
+        this.tableNotifier = Observable.getNotifier(this.table);
+        this.tableNotifier.subscribe(this, 'rowElements');
+    }
+
+    public handleChange(source: unknown, args: unknown): void {
+        if (
+            source === this.table
+            && args === 'rowElements'
+            && this.hasRowOrCellFocusType()
+            && this.inNavigationMode
+        ) {
+            this.focusCurrentRow(false);
+        }
+    }
+
+    public onVirtualizerChange(): void {
+        if (this.hasRowOrCellFocusType() && this.inNavigationMode) {
+            this.focusCurrentRow(false);
+        }
+    }
+
+    private readonly handleFocus = (event: FocusEvent): void => {
+        // User may have clicked elsewhere in the table (on an element not reflected in this.focusState). Update our focusState
```

Comment:
As indicated by the code comments, I'm not super satisfied with all of this logic being in the `focusin` handler, I think it makes it harder to reason about the various focus/unfocusing states of the table.

If we only had to worry about keyboard navigation by itself (without mouse interactions), the responsibilities of this class would be pretty simple - basically just a state machine: `current focus state + a key press => new focus state`.  
But in practice, the user can click anywhere on the table (like interacting with an action menu), and then we still want keyboard navigation to behave in a sane way.

Here's 2 action menu scenarios that behave differently, that led to the current logic in this function:
- Table is not focused at all, then the user clicks a cell action menu button. Result: first the `MenuButton` is the active element, then a `MenuItem` is active (once the menu is open)
- Table is focused (via Tab), the 1st row is focused (Down Arrow), then the user clicks the action menu button on some other row (not the 1st). Result: the `MenuItem` is active. I didn't see any focus events get triggered with the actual MenuButton in this scenario, which is why I have code going from a `MenuItem` back to the linked action menu.

This function also handles setting the appropriate element in the table on initial focus, so there's a fair amount going on in this.

I think there might be better ways to handle all of this, so I'm looking for suggestions on different approaches here too.

Reply:
A simple improvement might be to do a "clean code" pass and see if you can move each of the `if` blocks into their own well-named functions.

Another idea is to make `focusState` into a class rather than an object with individual fields being mutated. That might help make the state transitions more explicit. This might look something like:
1. make all the fields of the class `readonly`/getters from the outside
2. add public methods to set state in batches. e.g. 
```
focusCellActionMenu(rowIndex, columnIndex) {
   this.focusType = TableFocusType.cellActionMenu;
   this.rowIndex = rowIndex;
   this.columnIndex = columnIndex;
}
```
3. see if those public methods can enable this class to be even more explicit about state transitions. Maybe each one can be called from a navigation manager method that also updates the navigation manager's state.

I only looked at a couple of state transitions so I'm not sure how much this would help, might require prototyping one or two states in a branch to see if we like it better.

Reply:
Ah, I see Meyer made similar suggestions in https://github.com/ni/nimble/pull/2070/files#r1591068527